### PR TITLE
ServerReplicateActors Rate Limiting

### PR DESCRIPTION
**Pull request by joshuahuburn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description

Following the investigation into ServerReplicateActors we have decided to implement a basic rate limiting system until further optimisations can be made.

By setting a config variable in `DefaultEngine.ini` we can now reduce the amount of actors replicated per tick. Doing so significantly increases performance.
```
[/Script/SpatialGDK.SpatialNetDriver]
NetConnectionClassName="/Script/SpatialGDK.SpatialNetConnection"
ActorReplicationRateLimit=100
```

To do this we needed to fix actor prioritisation since this was not working for the 'fake' `SpatialConnection` which is actually the server-workers connection to the spatial runtime.
We now 'borrow' all player controllers when prioritising actors for replication to get an accurate list of `NetViewers`, using these we can calculate actor priority based on position and last time since replication.
This ensures that even with a rate limit, all actors will be replicated eventually thanks to the increasing priority.

Also added some Stats for useful debugging info.

#### Tests
On a 1000 actor test map, `ServerReplicateActors` was replicating 1000 actors per tick taking around 16ms.
With rate limiting, setting this to 100 actors per tick gives us 3ms.
We can set this value to roughly what native Unreal does which is 10~ actors per tick and visually the game performs the exact same as native.

#### Documentation
Documentation should follow. 
#### Primary reviewers
@davedolben @girayimprobable

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 1]
```diff
@@ -26,6 +26,10 @@
 
 DEFINE_LOG_CATEGORY(LogSpatialActorChannel);
 
+DECLARE_CYCLE_STAT(TEXT("SpatialActorChannel ~ ReplicateActor"), STAT_SpatialActorChannelReplicateActor, STATGROUP_SpatialNetDriver);
+DECLARE_CYCLE_STAT(TEXT("SpatialActorChannel ~ UpdateSpatialPosition"), STAT_SpatialActorChannelUpdateSpatialPosition, STATGROUP_SpatialNetDriver);
+DECLARE_CYCLE_STAT(TEXT("SpatialActorChannel ~ ReplicateSubobject"), STAT_SpatialActorChannelReplicateSubobject, STATGROUP_SpatialNetDriver);
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: No need to do the `SpatialActorChannel ~` bit imo. It's obvious from `STAT` name if anyone wants to disambiguate.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 2]
```diff
@@ -32,6 +32,9 @@
 
 DEFINE_LOG_CATEGORY(LogSpatialOSNetDriver);
 
+DECLARE_CYCLE_STAT(TEXT("SpatialNetDriver ~ ServerReplicateActors"), STAT_SpatialServerReplicateActors, STATGROUP_SpatialNetDriver);
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Refer to comments above for stat text.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h [hunk 3]
```diff
@@ -121,6 +124,9 @@ class SPATIALGDK_API USpatialNetDriver : public UIpNetDriver
 	UPROPERTY()
 	USnapshotManager* SnapshotManager;
 
+	UPROPERTY(Config)
+	int32 ActorReplicationRateLimit;
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h]: Add a comment re: what this actually does.

**Comment by girayimprobable** [comment]
This is user facing and needs to be documented. Since it's a bit in flux, it makes sense to make documentation a separate task. Please create one and label it with `documentation`

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 1 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Just copied the style from Unreal where I found the code originally, however they are inconsistent in the code base anyway so don't mind making this change.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h [hunk 4]
```diff
@@ -121,6 +124,13 @@ class SPATIALGDK_API USpatialNetDriver : public UIpNetDriver
 	UPROPERTY()
 	USnapshotManager* SnapshotManager;
 
+	// Setting ActorReplicationRateLimit in DefaultEngine.ini will limit the number of actors which are replicated per tick to the number specified.
```
Comment (davedolben) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h]: Setting ActorReplicationRateLimit in the `/Script/SpatialNetDriver` section, or whatever the proper section name is.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 5]
```diff
@@ -671,6 +685,11 @@ int32 USpatialNetDriver::ServerReplicateActors_ProcessPrioritizedActors(UNetConn
 	int32 ActorUpdatesThisConnectionSent = 0;
 	int32 FinalRelevantCount = 0;
 
+	// SpatialGDK - Actor replication rate limiting based on config value.
+	int32 RateLimit = ActorReplicationRateLimit > 0 ? ActorReplicationRateLimit : INT32_MAX;
+	RateLimit = FinalSortedCount < RateLimit ? FinalSortedCount : RateLimit;
+	int32 FinalReplicatedCount = RateLimit;
```
[the remark this answers is not in the document]
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Good idea, sorted.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 6]
```diff
@@ -765,47 +787,49 @@ int32 USpatialNetDriver::ServerReplicateActors_ProcessPrioritizedActors(UNetConn
...
-							LastRelevantActors.Add(Actor);
-						}
 
-						if (Channel->ReplicateActor())
+						// if the channel isn't saturated
+						if (Channel->IsNetReady(0))
 						{
-							ActorUpdatesThisConnectionSent++;
+							// replicate the actor
+							UE_LOG(LogNetTraffic, Log, TEXT("- Replicate %s. %d"), *Actor->GetName(), PriorityActors[j]->Priority);
```
[the remark this answers is not in the document]
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Ah this is just an existing log which is showing as changed thanks to whitespace ^^ This is the same in `NetDriver.cpp`

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 7]
```diff
@@ -708,26 +727,30 @@ int32 USpatialNetDriver::ServerReplicateActors_ProcessPrioritizedActors(UNetConn
...
 				{
 					LastNonRelevantActors.Add(Actor);
 				}
 			}
 
+			// SpatialGDK - We will only replicate the highest priority actors up the the rate limit.
+			// Actors not replicated this frame will have their priority increased based on the time since the last replicated.
+			if (RateLimit > 0)
+			{
+				bIsRelevant = true;
```
[the remark this answers is not in the document]
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: I see what you're saying, we kinda need to use bIsRelevant here though I think.
We're replacing the code 
```
					if ( IsActorRelevantToConnection( Actor, ConnectionViewers ) )
					{
						bIsRelevant = true;
					}
```

with our rate limiting code. To keep consistent with Unreal using the `const bool bIsRecentlyRelevant = bIsRelevant || (Channel && Time - Channel->RelevantTime < RelevantTimeout);` is required for setting up an actor channel if one doesn't exist. 
We could change the function a lot more to make it look nicer but I'm veering on the side of keep it as close to native as possible. Thoughts?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 8]
```diff
@@ -926,6 +930,26 @@ int32 USpatialNetDriver::ServerReplicateActors(float DeltaSeconds)
 					}
 				}
 			}
+			else
+			{
+				// Not having a ViewTarget implies this must be the fake spatial connection and thus borrow the player controllers from other connections.
+				for (int j = 0; j < ClientConnections.Num(); j++)
+				{
+					USpatialNetConnection* OtherSpatialConnection = Cast<USpatialNetConnection>(ClientConnections[j]);
```
[the remark this answers is not in the document]
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Based on the what the previous code was doing I'm gonna add a check, if it hasn't failed yet then it shouldn't fail now.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 9]
```diff
@@ -454,9 +457,20 @@ void USpatialNetDriver::NotifyActorDestroyed(AActor* ThisActor, bool IsSeamlessT
 // Returns true if this actor should replicate to *any* of the passed in connections
 static FORCEINLINE_DEBUGGABLE bool IsActorRelevantToConnection(const AActor* Actor, const TArray<FNetViewer>& ConnectionViewers)
 {
-	//SpatialGDK: Currently we're just returning true as a worker replicates all the known actors in our design.
+	// SpatialGDK: Currently we're just returning true as a worker replicates all the known actors in our design.
 	// We might make some exceptions in the future, so keeping this function.
 	return true;
+
+	// TODO: UNR-837 Start using IsNetRelevantFor again for relevancy checks rather than returning true.
```
[the remark this answers is not in the document]
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Sure I don't mind.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 10]
```diff
@@ -765,47 +776,49 @@ int32 USpatialNetDriver::ServerReplicateActors_ProcessPrioritizedActors(UNetConn
...
+							if (Channel->ReplicateActor())
+							{
+								ActorUpdatesThisConnectionSent++;
+								if (DebugRelevantActors)
+								{
+									LastSentActors.Add(Actor);
+								}
+
+								// Calculate min delta (max rate actor will update), and max delta (slowest rate actor will update)
+								const float MinOptimalDelta = 1.0f / Actor->NetUpdateFrequency;
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Is this protected against divide by zero?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 11]
```diff
@@ -671,6 +685,11 @@ int32 USpatialNetDriver::ServerReplicateActors_ProcessPrioritizedActors(UNetConn
 	int32 ActorUpdatesThisConnectionSent = 0;
 	int32 FinalRelevantCount = 0;
 
+	// SpatialGDK - Actor replication rate limiting based on config value.
+	int32 RateLimit = ActorReplicationRateLimit > 0 ? ActorReplicationRateLimit : INT32_MAX;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]:
```suggestion
	int32 RateLimit = (ActorReplicationRateLimit > 0) ? ActorReplicationRateLimit : INT32_MAX;
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 12]
```diff
@@ -671,6 +685,11 @@ int32 USpatialNetDriver::ServerReplicateActors_ProcessPrioritizedActors(UNetConn
...
+	RateLimit = FinalSortedCount < RateLimit ? FinalSortedCount : RateLimit;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]:
```suggestion
	RateLimit = (FinalSortedCount < RateLimit) ? FinalSortedCount : RateLimit;
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 13]
```diff
@@ -708,26 +727,30 @@ int32 USpatialNetDriver::ServerReplicateActors_ProcessPrioritizedActors(UNetConn
 		// Normal actor replication
 		USpatialActorChannel* Channel = Cast<USpatialActorChannel>(PriorityActors[j]->Channel);
 		UE_LOG(LogNetTraffic, Log, TEXT(" Maybe Replicate %s"), *PriorityActors[j]->ActorInfo->Actor->GetName());
-		if (!Channel || Channel->Actor) //make sure didn't just close this channel
+		if (!Channel || Channel->Actor) // make sure didn't just close this channel
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]:
```suggestion
		if (Channel == nullptr || Channel->Actor) // make sure didn't just close this channel
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 10 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: This is just a whitespace change and so should be fine.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 14]
```diff
@@ -765,47 +776,49 @@ int32 USpatialNetDriver::ServerReplicateActors_ProcessPrioritizedActors(UNetConn
...
-					// second check for channel saturation
-					if (!InConnection->IsNetReady(0))
-					{
-						// We can bail out now since this connection is saturated, we'll return how far we got though
-						return j;
+						// second check for channel saturation
+						if (!InConnection->IsNetReady(0))
+						{
+							// We can bail out now since this connection is saturated, we'll return how far we got though
+							return j;
```
[the remark this answers is not in the document]
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Dat.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h [hunk 15]
```diff
@@ -121,6 +124,13 @@ class SPATIALGDK_API USpatialNetDriver : public UIpNetDriver
 	UPROPERTY()
 	USnapshotManager* SnapshotManager;
 
+	// Setting ActorReplicationRateLimit in DefaultEngine.ini `[/Script/SpatialGDK.SpatialNetDriver]` section will limit the number of actors which are replicated per tick to the number specified.
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h]:
```suggestion
	// Limit the number of actors which are replicated per tick to the number specified.
```

This doesn't have to be `DefaultEngine.ini`, can be at different layers of the Engine ini stack and the section should be obvious to any UE4 developer.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h [hunk 16]
```diff
@@ -121,6 +124,13 @@ class SPATIALGDK_API USpatialNetDriver : public UIpNetDriver
...
+	// This acts as a hard limit to the number of actors per frame but nothing else. It's recommended to set this value to around 100~ (experimentation recommended).
+	// If not set spatial will replicate every actor per frame (unbounded) and so large worlds will experience slowdown server-side and client-side.
```
Comment (girayimprobable) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h]:
```suggestion
	// If not set SpatialOS will replicate every actor per frame (unbounded) and so large worlds will experience slowdown server-side and client-side.
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 17]
```diff
@@ -26,6 +26,10 @@
 
 DEFINE_LOG_CATEGORY(LogSpatialActorChannel);
 
+DECLARE_CYCLE_STAT(TEXT("ReplicateActor"), STAT_SpatialActorChannelReplicateActor, STATGROUP_SpatialNet);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Should we prefix all our added performance tags with "Spatial_" ?

**Pull request merged**